Realm content editing tools

JulianKniephoff opened the pull request:
This will close #290.

This PR will implement a first version of the realm content editor. It allows you to

* delete blocks from a realm,
* change the order of blocks within a realm,
* edit the properties of a block,
* and insert a new block anywhere in the list.

Some notes about design and implementation decisions, mostly deficiencies we accept for now in the interest of a looming deadline: 😐

* We will (for now) completely *ignore* problems arising from concurrently editing the realm/blocks, in the sense that, sure, the site should not crash, but any other kind of weird behavior resulting from it is fine.
  * This is especially true for concurrent edits from different clients. Within a single tab, we will (try our best to) prevent concurrent edits by disabling the corresponding controls.
    * **Note**: In its current version this is only prevented for standard "visual browsers." Assistive devices will probably not pick up on it since we are using visual trickery to achieve it, not any kind of semantic means.
* Reordering blocks will just be implemented using "up and down arrows" like for the child realms. Specifically, there will be no drag&drop functionality (for now). We need this for a11y purposes anyway, and it is sufficient for the first alpha release.
* We currently can't prevent the user from losing data when changing the page using our internal routing. Only refreshes and navigation to foreign pages will trigger a warning when a block is in edit mode (including new blocks). Basically anything that forces our page to actually unload.
  * Note that this also does **not** include the "back" button!
* There is no preview mode for editing blocks. If you want a preview, you have to save. :shrug: 
* In general the UX/UI of the page can probably be improved a lot, still.
  * The edit modes are very bare bones.
  * Save buttons could be disabled when the edited block is invalid.
  * Maybe errors could be highlighted better; like we could scroll them into view?
    * In general, when to show an error, where, and for how long is a hard question that I don't claim to have found a good answer for, here.
  * In general, animations might help make things look a bit less complex. :shrug:

## About the review

The commit history is reasonably clean and might contain useful information for a review, but I think it's best to start by reviewing the changes in full. Lot's of stuff gets moved around and changed over the course of the history. There are some things not directly related to this feature, and I tired keep at least those to their own commits, so maybe a quick glance over the history might be useful to pick those out and consequently ignore them in the "big review." :shrug:

LukasKalbertodt commented on frontend/src/routes/manage/Realm/Content/Edit/MoveButtons.tsx:
```diff
@@ -0,0 +1,77 @@
+    onError,
+}) => {
+    const { t } = useTranslation();
+
+
+    const { id: realmId, blocks } = useFragment(graphql`
+        fragment MoveButtonsData on Realm {
+            id
+            # We need this list only for the length,
+            # but we have to query *something* from it.
+            blocks { id }
+        }
+    `, realm);
+
+
```
Two empty lines again

JulianKniephoff replied:
Yes, see above. x)

LukasKalbertodt commented on frontend/src/routes/manage/Realm/Content/Edit/MoveButtons.tsx:
```diff
@@ -0,0 +1,77 @@
+    const { id: realmId, blocks } = useFragment(graphql`
+        fragment MoveButtonsData on Realm {
+            id
+            # We need this list only for the length,
+            # but we have to query *something* from it.
+            blocks { id }
```
I was already thinking whether passing the values directly (instead of using fragments) would be easier in this whole "edit content blocks" thing. Dunno :shrug:

JulianKniephoff replied:
Yes maybe, but I'm not gonna make that refactor now anymore unless you ask really really really really nicely. :stuck_out_tongue_winking_eye: 

But yes, trying to consistently use the Relay idea in these components was a major pain. I would even go so far, though, as to say that querying everything on top and just passing it through wouldn't be any better. You would still have to decide what to pass every time, and it is the fact that this decision was often wrong that was giving me a lot of trouble. In the future I think I would just pass the whole damn object around or even make a context. Yes, something something coupling, but it's not like the individual subcomponents of this page will see much reuse ... They are highly coupled anyway. :shrug:

LukasKalbertodt replied:
Yeah mh interesting. Probably a discussion for another time, but you are right: not in this PR.

LukasKalbertodt commented on frontend/src/routes/manage/Realm/Content/Edit/EditMode/index.tsx:
```diff
@@ -0,0 +1,174 @@
+
+    const modalRef = useRef<ConfirmationModalHandle>(null);
+
+    return <ButtonGroup css={{ marginTop: -24 }}>
```
Magic number? I guess absolute positioning and `top: 0` is more idiomatic?

JulianKniephoff replied:
Nop:

![image](https://user-images.githubusercontent.com/123272/150397106-754aeb1c-9494-4c59-bdb7-35ff696ca5b4.png)

Also it's not the only magic number in our CSS. :wink:

LukasKalbertodt replied:
Ye sure, css stuff is full of magic numers. And by `top:0` I of course also meant `right: 0`. But yeah, nevermind.

LukasKalbertodt commented on frontend/src/routes/manage/Realm/Content/Edit/EditMode/index.tsx:
```diff
@@ -0,0 +1,174 @@
+            <FiCheck />
+        </Button>
+        <ConfirmationModal
+            title={t("manage.are-you-sure")}
```
I've seen this line lots of times already. Maybe make that the default value of `ConfirmationModal` or sth...

JulianKniephoff replied:
Although that would make `ConfirmationModal` depend on the `i18n` stuff, which it currently does not. I mean that's a weak argument, admittedly, but that also means it can't just be a default argument, because we don't have access to the translation function at definition time. Or we could pass the key? :thinking: 

Or we just hard code it in `ConfirmationModal`.

Opinions? :sweat_smile:

LukasKalbertodt replied:
Hard code or make the prop optional and do `title ?? t("...)"` somewhere inside that component.

JulianKniephoff replied:
Done the latter in f3ed39a.

LukasKalbertodt commented on frontend/src/routes/manage/Realm/Content/Edit/EditMode/Text.tsx:
```diff
@@ -0,0 +1,98 @@
+                    : bug("not a text block");
+
+                save({
+                    variables: {
+                        id,
+                        set,
+                    },
```
This should comfortably fit in one line, no?

LukasKalbertodt commented on frontend/src/routes/manage/Realm/Content/Edit/EditMode/Text.tsx:
```diff
@@ -0,0 +1,98 @@
+                    : bug("not a text block");
+
+                create({
+                    variables: {
+                        realm,
+                        index,
+                        block,
+                    },
```
Also still one line

LukasKalbertodt commented on frontend/src/routes/manage/Realm/Content/Edit/EditMode/Series.tsx:
```diff
@@ -0,0 +1,199 @@
+
+        const { series: allSeries } = useFragment(graphql`
+            fragment SeriesEditModeSeriesData on Query {
+                series {
+                    id
+                    title
+                }
```
One line?

LukasKalbertodt commented on frontend/src/routes/manage/Realm/Content/Edit/EditMode/Series.tsx:
```diff
@@ -0,0 +1,199 @@
+            fragment SeriesEditModeBlockData on SeriesBlock {
+                order
+                layout
+                series {
+                    id
+                }
```
One line?

LukasKalbertodt reviewed (changes requested):
Overall nice work. Really glad to have this feature now. Huge improvement for the project as a whole. I left quite a few comments though. Many of them not really critical. And yeah, been some time since I last reviewed a PR of this size.

As a general note, I am not really a fan of using the relay store like that. I would only use relay for GraphQL operations and not insert our own dummy data. However, I am sure the alternative solution would have its fair share of problems *and* I did not implement this feature, so you have more experience to judge this. In any case, we won't change this now anymore.

From my testing, I also found these things:

----

**Bugs & stuff**
- When adding a second series block without reloading the page (and in a few more scenarios), I get:
  ```
  Uncaught Invariant Violation: RelayRecordSourceMutator#create(): Cannot create a record with id `clNOSERIES`, this record already exists.
  ```
- Series `<select>` should get `max-width: 100%` and more `input` like styling (e.g. `border-radius: 4px`)
- When clicking on an arrow to move a block, during the operation, the UI is *not* disabled so with slow internet it's quite easy to start two swap operations at the same time. Maybe just grey out everything?

**Suggestions**
- When being in block-edit-mode, everything else is greyed out. I would make this effect stronger, i.e. lower opacity.
- I intuitively expect a save button at the bottom of the block edit thing. I think I would only leave the X on the top right and put a normal save/add button at the bottom.
- Clicking on "add block" and immediately on "discard" (without having specified any data), it still warns. That shouldn't happen. At least for text block, where it's probably simple to check if sth changed/was added.
- I am still not a big fan of the adding buttons :/
- There should be a link back to the actual page on the "edit content" page.
- The error box for commit errors does not have margin-bottom and directly touches the affected block. Maybe add at least a little margin? Otherwise I think the error is totally sufficient.


**For the future, but no in this PR**
- Eventuelly, the series select needs a search/filter functionality.
- Layout option for series block should have visual indicators in the future, but nvm now
- Be more consistent with breadcrumbs and maybe put them on this page too?
- In the series select, we might want to show more than the title, as there are a lot of duplicate titles.

**Random thoughts**
- Maybe after adding a page, one should get forwarded to the edit mode instead of the empty page?
- Maybe empty pages should render "Empty page" or sth?

JulianKniephoff commented on frontend/src/routes/manage/Realm/DangerZone.tsx:
```diff
@@ -209,25 +204,21 @@ const RemoveRealm: React.FC<InnerProps> = ({ realm }) => {
         </p>
         <div css={{ marginTop: 32, textAlign: "center" }}>
-            <Button kind="danger" onClick={() => setModalActive(true)}>
+            <Button kind="danger" onClick={() => modalRef.current?.open()}>
```
Regarding the first point: No. Optional chaining short-circuits:

```javascript
({}).foo?.bar()
```

is just `undefined`.

`!` is currently not allowed by our ESLint rules. Because of that I have been going with things like

```javascript
const theRealThing = theThing ?? bug("the thing is not real");
```

elsewhere. I could do something like this here as well. The advantage is that you are forced to make the failure explicit. Most of the time if you would use `!` wrongly you would get an error anyway, like in this case, because you try to access stuff on `undefined`, but in general one can imagine cases where constructs with `bug` fail loudly while `!` would just lead to weird behavior.

What do you think? :thinking:

LukasKalbertodt replied:
Yeah right, `!` is not a runtime check at all right? So ye, `?? bug` is better in that regard. In this case, I don't care. The current version is fine too.

JulianKniephoff replied:
Exactly, `!` is just stripped out during compilation. It's just to appease the type checker.

The problem with the `bug` approach in this particular case is that you have to spell it out every time. You can't memoize it on the top level of a component, precisely because it's a ref. What do you think about a helper function, maybe in `err`, a la

```javascript
const currentRef = ref => ref.current ?? bug("unbound ref");
```

?

LukasKalbertodt replied:
Yeah why not. Sounds like a use case that happens often enough. "Ref should be bound in handler" or sth. Ye.

JulianKniephoff replied:
Ok, it's done in e072ca3. I also transitioned some of your code to it. Just the obvious cases, though. You might want to double check me. :shrug:

JulianKniephoff commented on frontend/src/ui/Modal.tsx:
```diff
@@ -1,50 +1,163 @@
+            return <Modal title={title} closable={!inFlight} ref={modalRef}>
+                {children}
+                <form onSubmit={onSubmitWrapper} css={{ marginTop: 32, textAlign: "center" }}>
+                    <Button autoFocus={true} disabled={inFlight} type="submit" kind="danger">
```
Hmm, good point. :thinking: 

It's of course a million times more convenient, especially for devs during testing. x)

And I don't know whether you really accidentally push the delete button and then press enter. :thinking:

LukasKalbertodt replied:
I don't think I would value DX over the potential UX disaster that comes from this. So I think I would like to not auto focus.

LukasKalbertodt replied:
I don't think I would value DX over the potential UX disaster that comes from this. So I think I would like to not auto focus.

JulianKniephoff replied:
Alright, will get rid of it. :shrug:

JulianKniephoff replied:
Done in d4e8a04.

JulianKniephoff commented on backend/src/api/mutation.rs:
```diff
@@ -41,7 +52,59 @@ impl Mutation {
+        BlockValue::add_series(realm, index, block, context).await
+    }
+
+    /// Swap two blocks.
+    ///
+    /// This is a less flexible but potentially cheaper alternative
+    /// to `orderBlocks`.
+    async fn swap_blocks_by_id(id_1: Id, id_2: Id, context: &Context) -> ApiResult<Realm> {
+        BlockValue::swap_by_id(id_1, id_2, context).await
+    }
```
Yeah I wasn't sure whether we might not want to keep it around, though? As we discussed, it's unclear which version (if any) of the swapping APIs is really preferable. :thinking: 

I could of course add another commit removing it, so it's at least in the history, still. What do you think?

LukasKalbertodt replied:
If we don't currently use it, I would remove it in another commit (to keep it in the history). And I guess swap by index is fine for now. So ye.

JulianKniephoff replied:
See 57293a3.

JulianKniephoff commented on backend/src/api/model/block/mutations.rs:
```diff
@@ -0,0 +1,355 @@
+        Ok(())
+    }
+
+    pub(crate) async fn swap_by_id(id_1: Id, id_2: Id, context: &Context) -> ApiResult<Realm> {
```
Yeah, see above.

JulianKniephoff replied:
See 57293a3.

JulianKniephoff commented on backend/src/api/model/block/mutations.rs:
```diff
@@ -0,0 +1,355 @@
+        context: &Context
+    ) -> ApiResult<Realm> {
+
+        if index_1 == index_2 {
+            return Realm::load_by_id(realm, context)
+                .await?
+                .ok_or_else(|| invalid_input!("`realm` is not a valid realm"));
+        }
```
Well, I didn't really see it as an error. You can swap an element of a list with itself, it's just a no-op.

I can see the argument that you would probably never want to do that, so if you do, it's most likely a mistake, though. :thinking:

LukasKalbertodt replied:
Ah yeah, ok yeah it's fine as is.

JulianKniephoff commented on backend/src/api/model/block/mutations.rs:
```diff
@@ -0,0 +1,355 @@
+        let realm_stream = context.db(context.require_moderator()?)
+            .query_raw(
+                &format!(
+                    "update blocks
+                        set index = updates.new_index
+                        from realms, (values
+                            ($1::smallint, $2::smallint),
+                            ($2::smallint, $1::smallint)
+                        ) as updates(old_index, new_index), (
+                            select count(*) as count from blocks
+                            where realm_id = $3
+                        ) as count
+                        where realm_id = realms.id
+                        and realm_id = $3
+                        and blocks.index = updates.old_index
+                        and updates.new_index < count
+                        and updates.new_index >= 0
+                        returning {}",
```
Says the guy with the window function in his query! :stuck_out_tongue: 

I think I want to keep it, but sure I can add a comment.

Splitting the query could also make it more likely for the transaction to fail, right?

JulianKniephoff replied:
What about this? Can I keep my artisanal query, pretty please? :3

JulianKniephoff replied:
Comment added in 1602608.

JulianKniephoff commented on backend/src/api/model/block/mutations.rs:
```diff
@@ -0,0 +1,355 @@
+                &format!(
+                    "update blocks set
+                        title = case $2::boolean when true then $3 else title end,
+                        text_content = coalesce($4, title)
```
Yes regarding the `text_content`.

Regarding the second point, there is a difference. I wanted the ability to explicitly set the title to `null` again, since the frontend actually renders `""` titles differently, IIRC. However, that's not possible with `coalesce`. If you only ever update a field with `coalesce(argument, field)`, once the field was non-null once, that expression never evaluate to `null` again.

LukasKalbertodt replied:
Ah ok makes sense.

JulianKniephoff commented on frontend/src/i18n/locales/de.yaml:
```diff
@@ -194,6 +193,49 @@ manage:
+      heading-root: Inhalte der Startseite bearbeiten
+
+      add: Hier einen neuen Block einfügen
+      add-text: Hier einen neuen Text-Block einfügen
+      add-series: Hier einen neuen Serien-Block einfügen
+      add-video: Hier einen neuen Video-Block einfügen
```
Both are valid. I'm not sure what the prevalent style has been so far, but since you wrote most messages until now, and you think there shouldn't be a hyphen, it's probably what you suggested, right? :sweat_smile:

LukasKalbertodt replied:
"Fremdauthentifizierungsfehler" exists in this file :D 

I dunno. I think I would not use a hyphen here. But no blocker.

JulianKniephoff replied:
Done in 8e3601e.

JulianKniephoff commented on frontend/src/routes/manage/Realm/Content/index.tsx:
```diff
@@ -0,0 +1,158 @@
+                }
+            }
+        `,
+        data.realm as ContentManageRealmData$key,
```
Have you tried removing it? :wink: 

```
ERROR in src/routes/manage/Realm/Content/index.tsx:88:13
TS2339: Property 'name' does not exist on type 'unknown'.

ERROR in src/routes/manage/Realm/Content/index.tsx:88:19
TS2339: Property 'isRoot' does not exist on type 'unknown'.

ERROR in src/routes/manage/Realm/Content/index.tsx:88:40
TS2339: Property 'blocks' does not exist on type 'unknown'.

ERROR in src/routes/manage/Realm/Content/index.tsx:102:43
TS7006: Parameter 'block' implicitly has an 'any' type.

ERROR in src/routes/manage/Realm/Content/index.tsx:127:32
TS7006: Parameter 'block' implicitly has an 'any' type.

ERROR in src/routes/manage/Realm/Content/index.tsx:127:61
TS7006: Parameter 'block' implicitly has an 'any' type.

ERROR in src/routes/manage/Realm/Content/index.tsx:127:68
TS7006: Parameter 'index' implicitly has an 'any' type.

ERROR in src/routes/manage/Realm/Content/index.tsx:129:51
TS2322: Type 'unknown' is not assignable to type 'AddButtonsRealmData$key'.

ERROR in src/routes/manage/Realm/Content/index.tsx:132:30
TS2322: Type '{ onCommit: () => void; onCompleted: () => void; onError: () => void; realm: unknown; index: any; }' is not assignable to type 'Props'.
  Types of property 'realm' are incompatible.
    Type 'unknown' is not assignable to type 'BlockRealmData$key'.

ERROR in src/routes/manage/Realm/Content/index.tsx:141:51
TS2322: Type 'unknown' is not assignable to type 'AddButtonsRealmData$key'.
```

JulianKniephoff replied:
But granted, thinking about this again seems a bit fishy. :thinking:

JulianKniephoff replied:
What do we want to do about this now?

LukasKalbertodt replied:
Nothing :P

JulianKniephoff commented on frontend/src/routes/manage/Realm/Content/index.tsx:
```diff
@@ -0,0 +1,158 @@
+                        <div css={block.editMode && !inFlight ? { zIndex: 2 } : {}}>
+                            <EditBlock
+                                {...{ realm, index }}
+                                onCommit={onCommit}
```
Symmetry with the other two handlers.

JulianKniephoff commented on frontend/src/routes/manage/Realm/Content/Edit/RemoveButton.tsx:
```diff
@@ -0,0 +1,87 @@
+
+export const RemoveButton: React.FC<Props> = ({ block: blockRef, onConfirm }) => {
+    const { t } = useTranslation();
+
+
+    const block = useFragment(graphql`
+        fragment RemoveButtonData on Block {
+            id
+        }
+    `, blockRef);
+
+
```
Well, both the stuff above and below it have a different "topic." It's either both, or just one line on both sides (and everywhere else in that function). Like, double-newline is used to group things that potentially have "subgroups" separated by single newlines. If there was just a single newline on either end, it would mark this snippet as belonging to the corresponding group, which is not what I intended to express.

LukasKalbertodt replied:
:shrug:

JulianKniephoff commented on backend/src/api/model/series.rs:
```diff
@@ -32,6 +34,24 @@ impl Series {
+        let series = context.db(context.require_moderator()?)
+            .query_raw(
+                &format!(
+                    "select {} from series
+                    order by title",
```
Yes, nice wrapper! We'll see who's first, though. :wink:

JulianKniephoff replied:
Well, I was! :stuck_out_tongue:

JulianKniephoff commented on backend/src/api/query.rs:
```diff
@@ -37,6 +47,11 @@ impl Query {
         Event::load_by_id(id, context).await
     }
 
+    /// Returns a (potentially paginated) list of all series the current user has access to
+    async fn series(context: &Context) -> ApiResult<Vec<Series>> {
+        Series::load(context).await
+    }
```
Yeah, that's BS. I wrote that very early, before it was clear that we won't put much effort into what series to list and how to list them. The part about the access is a lie, too. Will replace with basically the first part without the parens.

JulianKniephoff commented on frontend/src/routes/manage/Realm/DangerZone.tsx:
```diff
@@ -209,25 +204,21 @@ const RemoveRealm: React.FC<InnerProps> = ({ realm }) => {
-        )}
+        <ConfirmationModal
+            title={t("manage.are-you-sure")}
+            {...{ buttonContent }}
+            onSubmit={remove}
+            ref={modalRef}
```
I roughly order fields by "relevance" regardless of how I pass them. If you insist on this style, I think I would rather get rid of the spread.

Note also that this can never be a hard and fast rule. There is a semantic difference. If props are shared between the key-value pairs and the spread, the order determines which one wins. And I can imagine use cases for both to win. (Basically: Are the attributes just a set of defaults overridable with the user-passed spread or do you want to explicitly prevent the spread overriding some props.)

LukasKalbertodt replied:
Ah ye ok. Then let's leave it like that.

JulianKniephoff commented on frontend/src/routes/manage/Realm/DangerZone.tsx:
```diff
@@ -190,10 +186,9 @@ const RemoveRealm: React.FC<InnerProps> = ({ realm }) => {
             onError: error => {
                 const failedAction = t("manage.realm.danger-zone.delete.failed");
-                setCommitError(displayCommitError(error, failedAction));
+                modalRef.current?.reportError(displayCommitError(error, failedAction));
```
See above

JulianKniephoff commented on frontend/src/ui/Modal.tsx:
```diff
@@ -1,50 +1,163 @@
+                open: () => {
+                    setInFlight(false);
+                    setError(undefined);
+                    modalRef.current?.open();
```
See above.

JulianKniephoff commented on frontend/src/ui/Blocks/index.tsx:
```diff
@@ -0,0 +1,80 @@
+    `, realmRef);
+    const { blocks } = realm;
+
+    return <div css={{
+        display: "flex",
+        flexDirection: "column",
+        rowGap: 32,
+    }}>
+        {blocks.map(
+            block => <Block key={block.id} realm={realm} block={block} />,
+        )}
+    </div>;
```
I don't mind it, but I also don't mind changing it. :shrug:

JulianKniephoff commented on backend/src/api/mutation.rs:
```diff
@@ -41,7 +52,59 @@ impl Mutation {
+    /// Swap two blocks.
+    ///
+    /// This is a less flexible but potentially cheaper alternative
+    /// to `orderBlocks`.
```
True. :sweat_smile:

JulianKniephoff commented on backend/src/api/mutation.rs:
```diff
@@ -41,7 +52,59 @@ impl Mutation {
+    /// to `orderBlocks`.
+    async fn swap_blocks_by_index(
+        realm: Id,
+        index_1: i32,
+        index_2: i32,
```
Hmm I think I like that. :thinking:

JulianKniephoff commented on backend/src/api/model/block/mutations.rs:
```diff
@@ -0,0 +1,355 @@
+impl BlockValue {
+    pub(crate) async fn add_text(
+        realm: Id,
+        index: i32,
```
After or before is the wrong point of view, I think. It inserts **at** the index, as in: the inserted block will have that index. (Which in your worldview works out to be the "before" case, but that's incidental. :shrug: )

But sure, I can add a comment to that effect.

JulianKniephoff commented on backend/src/api/model/block/mutations.rs:
```diff
@@ -0,0 +1,355 @@
+    }
+
+    async fn make_room_for_block(realm: Key, index: i16, context: &Context) -> ApiResult<()> {
+        let blocks: i64 = context.db
```
:shrug:

JulianKniephoff commented on backend/src/api/model/block/mutations.rs:
```diff
@@ -0,0 +1,355 @@
+            )
+            .await?;
+
+        Realm::load_by_key(realm, context)
+            .await?
+            .ok_or_else(|| invalid_input!("`realm` does not refer to a valid realm"))
```
I thought about a similar function; I don't remember the details, though. I decided against it, though, because there doesn't really seem to be a great abstraction here. Like, sometimes the check looks maybe slightly different because it's more convenient or sth. :shrug: 

Do you want me to add it or will you take it as is? :wink: 

Regarding the second point: I guess? :shrug: :sweat_smile: I need the key anyway, for the `make_room` function, and this way I can immediately return the realm. What's the advantage in your proposed change? :thinking:

LukasKalbertodt replied:
We can do the `try_load_by_key` or whatever in some of those famous "various improvements" PRs later :P 
And yeah nevermind the other stuff

JulianKniephoff commented on backend/src/api/model/block/mutations.rs:
```diff
@@ -0,0 +1,355 @@
+            .ok_or_else(|| invalid_input!("`realm` does not refer to a valid realm"))
+    }
+
+    pub(crate) async fn add_series(
```
Yeah I hope I got them all.

JulianKniephoff commented on backend/src/api/model/block/mutations.rs:
```diff
@@ -0,0 +1,355 @@
+                    &!set.title.is_implicit_null(),
+                    &set.title.some(),
+                    &set.series.map(
+                        |series| series.key_for(Id::SERIES_KIND)
+                            .ok_or_else(|| invalid_input!("`set.series` does not refer to a series"))
```
Yeah, again, I thought about something related, but decided against it, because it quickly proved to inflexible in my head. You can try coming up with something, though. :shrug:

LukasKalbertodt replied:
I added it to my own TODO list, lets see if I can come up with something.

JulianKniephoff commented on backend/src/api/model/block/mutations.rs:
```diff
@@ -0,0 +1,355 @@
+            )
+            .await?;
+
+        let realm_key: Key = result.get(0);
```
Nope, that's private. :wink:

LukasKalbertodt replied:
Can't you... make it public? :D 
But ye nevermind, I don't care.

JulianKniephoff replied:
Sure, but I figured it was private for a reason. :sweat_smile: 

If you're cool with it, I will gladly make it public, since I agree: this is kind of an ugly workaround.

LukasKalbertodt replied:
Yes, public then please. Private is just the "default" in a sense, so there is not always a reason behind it.

JulianKniephoff replied:
Done in c6b640a.

JulianKniephoff commented on backend/src/api/model/block/mutations.rs:
```diff
@@ -0,0 +1,355 @@
+        // Fix indices after removed block
+        db
+            .execute(
+                "update blocks
+                    set index = index - 1
+                    where realm_id = $1
+                    and index > $2",
```
I think I got them all.

JulianKniephoff commented on frontend/src/i18n/locales/en.yaml:
```diff
@@ -186,6 +185,49 @@ manage:
+
+      removing-failed: Removing the block failed.
+      moving-failed: Moving the block failed.
+      saving-failed: Saving your changes to the block failed.
```
eh :shrug:

JulianKniephoff commented on frontend/src/routes/manage/Realm/Content/util.tsx:
```diff
@@ -0,0 +1,57 @@
+                },
+            },
+        }}
+        {...props}
```
It should pass `children` as well, right. But then I would say why destructure? Why not just remove the explicit mention of the children?

LukasKalbertodt replied:
Does that work? Can you pass the childrens prop with the attribute syntax? :shrug:

JulianKniephoff replied:
I think so.

LukasKalbertodt replied:
Well then, use whatever solution you like best :P

JulianKniephoff commented on frontend/src/routes/manage/Realm/Content/util.tsx:
```diff
@@ -0,0 +1,57 @@
+
+export const ButtonGroup: React.FC<ButtonGroupProps> = props => (
+    <div
+        {...props}
```
See above somewhere.

JulianKniephoff commented on frontend/src/routes/manage/Realm/Content/index.tsx:
```diff
@@ -0,0 +1,158 @@
+
+                        <div css={block.editMode && !inFlight ? { zIndex: 2 } : {}}>
+                            <EditBlock
+                                {...{ realm, index }}
```
See above.

JulianKniephoff commented:
OK, addressing the "global" comments:

> Overall nice work. Really glad to have this feature now. Huge improvement for the project as a whole. I left quite a few comments though. Many of them not really critical. And yeah, been some time since I last reviewed a PR of this size.

:blush: 

> As a general note, I am not really a fan of using the relay store like that. I would only use relay for GraphQL operations and not insert our own dummy data. However, I am sure the alternative solution would have its fair share of problems _and_ I did not implement this feature, so you have more experience to judge this. In any case, we won't change this now anymore.

I mean I only implemented the relay store approach. Thinking about the alternative, though, I also see lots of complications. We talked about some of them. We are talking about distributed state here. It's a complex problem. :shrug: I just prefer one source of truth.

> From my testing, I also found these things:
> 
> **Bugs & stuff**
> 
> * When adding a second series block without reloading the page (and in a few more scenarios), I get:
>   ```
>   Uncaught Invariant Violation: RelayRecordSourceMutator#create(): Cannot create a record with id `clNOSERIES`, this record already exists.
>   ```

Fixed in 0ca53e4.

> * Series `<select>` should get `max-width: 100%` and more `input` like styling (e.g. `border-radius: 4px`)

Done in 969e256.

> * When clicking on an arrow to move a block, during the operation, the UI is _not_ disabled so with slow internet it's quite easy to start two swap operations at the same time. Maybe just grey out everything?

Fixed in f756b85.

> **Suggestions**
> 
> * When being in block-edit-mode, everything else is greyed out. I would make this effect stronger, i.e. lower opacity.

How do you like `0.75` opacity (of the white overlay)?

> * I intuitively expect a save button at the bottom of the block edit thing. I think I would only leave the X on the top right and put a normal save/add button at the bottom.

At this point I would agree. However, I'm a bit hesitant to change it now, because in my view that would necessitate some other changes. You see, the fact that the submit button belongs to that button group had some structural consequences which I'm not happy with, but which I accepted to achieve this "consistency." If we change this now I would like to at least think about those again as well, and potentially revert them. It's not a huge change of course, but it might also not be "let me just push this in 5 minutes." Thoughts?

> * Clicking on "add block" and immediately on "discard" (without having specified any data), it still warns. That shouldn't happen. At least for text block, where it's probably simple to check if sth changed/was added.

I can very easily cover your case (see 6b10f96), but that still fails when you click into the form, for example, before canceling. That would require actual work. Dunno how important this is to us? :thinking: The actual "stupid" edge case is solved with this.

> * I am still not a big fan of the adding buttons :/

Yeah well, I don't think we will come to a good solution int he scope of this PR. :thinking: 

> * There should be a link back to the actual page on the "edit content" page.

How do you like 6b10f96?

> * The error box for commit errors does not have margin-bottom and directly touches the affected block. Maybe add at least a little margin? Otherwise I think the error is totally sufficient.

How do you like it now?

> **For the future, but no in this PR**
> 
> * Eventuelly, the series select needs a search/filter functionality.
> * Layout option for series block should have visual indicators in the future, but nvm now
> * Be more consistent with breadcrumbs and maybe put them on this page too?
> * In the series select, we might want to show more than the title, as there are a lot of duplicate titles.

Will add these and my own points to an issue once this is merged.

> **Random thoughts**
> 
> * Maybe after adding a page, one should get forwarded to the edit mode instead of the empty page?

Done in db8c5b1.

> * Maybe empty pages should render "Empty page" or sth?

:shrug: Where would you place that in relation to the add buttons? I think such details can wait until we have a "proper design," though. Whatever that means.

---

Alright, that should be everything. :sweat_smile:

LukasKalbertodt commented on frontend/src/ui/Modal.tsx:
```diff
@@ -1,50 +1,163 @@
+            children,
+        }, ref) => {
+            const [inFlight, setInFlight] = useState(false);
+            const [error, setError] = useState<JSX.Element | undefined>();
```
Wat. I thought `<type>?` was always valid in type position. Well then.

LukasKalbertodt commented:
> How do you like `0.75` opacity (of the white overlay)?

I do like it! Thx.

> At this point I would agree. However, I'm a bit hesitant to change it now, because in my view that would necessitate some other changes. You see, the fact that the submit button belongs to that button group had some structural consequences which I'm not happy with, but which I accepted to achieve this "consistency." If we change this now I would like to at least think about those again as well, and potentially revert them. It's not a huge change of course, but it might also not be "let me just push this in 5 minutes." Thoughts?

Mhhh. I guess we can merge it without this change, but I think we should tackle it soon (i.e. before feb alpha deadline). But we can ask Lars what he thinks, too. 

> I can very easily cover your case (see [6b10f96](https://github.com/elan-ev/tobira/commit/6b10f96f2886f84d4499ad0ea7724ff5a655eca1)), but that still fails when you click into the form, for example, before canceling. That would require actual work. Dunno how important this is to us?  The actual "stupid" edge case is solved with this.

Thanks. I guess it's good enough for now. Maybe add it to some list of potential improvements, but low prio.

> Yeah well, I don't think we will come to a good solution int he scope of this PR. thinking

True. Lets merge it and we can still test, ask Lisa and improve it later.


> > * There should be a link back to the actual page on the "edit content" page.
> 
> How do you like [6b10f96](https://github.com/elan-ev/tobira/commit/6b10f96f2886f84d4499ad0ea7724ff5a655eca1)?

Either you misunderstood or you linked the wrong commit?

> > * The error box for commit errors does not have margin-bottom and directly touches the affected block. Maybe add at least a little margin? Otherwise I think the error is totally sufficient.
> 
> How do you like it now?

:ok_hand: 

> > * Maybe empty pages should render "Empty page" or sth?
> 
> shrug Where would you place that in relation to the add buttons? I think such details can wait until we have a "proper design," though. Whatever that means.

Oh, I wasn't talking about the "edit content" page, but about the realm page itself. But ye, not this PR.

JulianKniephoff commented:
> Either you misunderstood or you linked the wrong commit?

The latter. I meant ac2a10d9690ab9464f4d69a1fbd83991136f3778. :sweat_smile:

JulianKniephoff commented:
> Still doesn't have the `max-width: 100%`.

And finally 7c3496d!

LukasKalbertodt reviewed (approved):
:tada:
